edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.Core.Twin
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Storage;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Shared;
+    using Microsoft.Extensions.Logging;
+    using Nito.AsyncEx;
+
+    class ReportedPropertiesStore : IReportedPropertiesStore
+    {
+        static readonly TimeSpan DefaultSyncFrequency = TimeSpan.FromSeconds(5);
+        readonly IEntityStore<string, TwinStoreEntity> twinStore;
+        readonly ICloudSync cloudSync;
+        readonly AsyncLockProvider<string> lockProvider = new AsyncLockProvider<string>(10);
+        readonly AsyncAutoResetEvent syncToCloudSignal = new AsyncAutoResetEvent(false);
+        readonly HashSet<string> syncToCloudClients = new HashSet<string>(StringComparer.Ordinal);
+        readonly object syncToCloudSetLock = new object();
+        readonly Task syncToCloudTask;
+        readonly TimeSpan syncFrequency;
+
+        public ReportedPropertiesStore(IEntityStore<string, TwinStoreEntity> twinStore, ICloudSync cloudSync, Option<TimeSpan> syncFrequency)
+        {
+            this.twinStore = twinStore;
+            this.cloudSync = cloudSync;
+            this.syncToCloudTask = this.SyncToCloud();
+            this.syncFrequency = syncFrequency.GetOrElse(DefaultSyncFrequency);
+        }
+
+        public async Task Update(string id, TwinCollection patch)
+        {
+            using (await this.lockProvider.GetLock(id).LockAsync())
+            {
+                Events.StoringReportedPropertiesInStore(id, patch);
+                await this.twinStore.PutOrUpdate(
+                    id,
+                    new TwinStoreEntity(patch),
+                    twinInfo =>
+                    {
+                        Events.UpdatingReportedPropertiesInStore(id, patch);
+                        TwinCollection updatedReportedProperties = twinInfo.ReportedPropertiesPatch
+                            .Map(reportedProperties => new TwinCollection(JsonEx.Merge(reportedProperties, patch, /*treatNullAsDelete*/ false)))
+                            .GetOrElse(() => patch);
+                        return new TwinStoreEntity(twinInfo.Twin, Option.Maybe(updatedReportedProperties));
+                    });
+            }
+        }
+
+        public void InitSyncToCloud(string id)
+        {
+            lock (this.syncToCloudSetLock)
+            {
+                this.syncToCloudClients.Add(id);
+            }
+
+            this.syncToCloudSignal.Set();
+        }
+
+        public async Task SyncToCloud(string id)
+        {
+            Events.SyncingReportedPropertiesToCloud(id);
+            Option<TwinStoreEntity> twinWithReportedProperties = (await this.twinStore.Get(id))
+                .Filter(t => t.ReportedPropertiesPatch.HasValue);
```

Comment:
The twin store contains the latest patch?

Reply:
It contains all the unapplied patches (which are applied on top of one another). 
Say if the device sends the following patches in quick succession (before any sync to IoTHub happens) - rp1, rp2, rp3 - then the store will contain the patches applied on top of one another - rp_store = { rp1 + rp2 + rp3 }